session/connectivity/status_storage.go

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (C) 2019 The "MysteriumNetwork/node" Authors.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package connectivity
+
+import (
+	"sync"
+	"time"
+
+	"github.com/mysteriumnetwork/node/identity"
+)
+
+// StatusStorage is responsible for status storage operations.
+type StatusStorage interface {
+	GetAllStatusEntries() []*StatusEntry
+	AddStatusEntry(msg *StatusEntry)
+}
+
+// StatusEntry describes status entry.
+type StatusEntry struct {
+	PeerID       identity.Identity
+	SessionID    string
+	StatusCode   StatusCode
+	Message      string
+	CreatedAtUTC time.Time
+}
+
+// NewStatusStorage returns new StatusStorage instance.
+func NewStatusStorage() StatusStorage {
+	return &statusStorage{}
+}
+
+type statusStorage struct {
+	entries    []*StatusEntry
+	entriesMux sync.RWMutex
+}
+
+func (s *statusStorage) GetAllStatusEntries() []*StatusEntry {
+	s.entriesMux.RLock()
+	defer s.entriesMux.RUnlock()
+
+	return s.entries
+}
+
+func (s *statusStorage) AddStatusEntry(msg *StatusEntry) {
+	s.entriesMux.Lock()
+	defer s.entriesMux.Unlock()
+
+	s.entries = append(s.entries, msg)
```

Comment:
Running node for a long time entries eventually may consume more memory as this is in memory storage. As I have CreatedAtUTC field I could remove older entires (1 month old etc.) when calling AddStatusEntry.

Reply:
as far as i'm aware, session storage emits events when a session is removed - you could subscribe to those and remove the relevant entries here as well.

Reply:
I think node runner may want to see if consumers can connect successfully thats why I also send success event.